compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala

```diff
@@ -0,0 +1,144 @@
+    def isNullableUnion(implicit ctx: Context): Boolean = {
+      assert(ctx.explicitNulls)
+      self.widenDealias.normNullableUnion match {
+        case OrType(_, rhs) => rhs.isNullType
+        case _ => false
+      }
+    }
+
+    /** Is self (after widening and dealiasing) a type of the form `T | JavaNull`? */
+    def isJavaNullableUnion(implicit ctx: Context): Boolean = {
+      assert(ctx.explicitNulls)
+      self.widenDealias.normNullableUnion match {
+        case OrType(_, rhs) => rhs.isJavaNullType
+        case _ => false
+      }
+    }
+
+    /** Is this type guaranteed not to have `null` as a value? */
+    final def isNotNull(implicit ctx: Context): Boolean = self match {
+      case tp: ConstantType => tp.value.value != null
+      case tp: ClassInfo => !tp.cls.isNullableClass && tp.cls != defn.NothingClass
+      case tp: TypeBounds => tp.lo.isNotNull
+      case tp: TypeProxy => tp.underlying.isNotNull
+      case AndType(tp1, tp2) => tp1.isNotNull || tp2.isNotNull
+      case OrType(tp1, tp2) => tp1.isNotNull && tp2.isNotNull
+      case _ => false
+    }
+
+    def maybeNullable(implicit ctx: Context): Type =
+      if (ctx.explicitNulls) OrType(self, defn.NullType) else self
+
+    /** Syntactically strips the nullability from this type.
+     *  If the normalized form (as per `normNullableUnion`) of this type is `T1 | ... | Tn-1 | Tn`,
+     *  and `Tn` references to `Null` (or `JavaNull`), then return `T1 | ... | Tn-1`.
+     *  If this type isn't (syntactically) nullable, then returns the type unchanged.
+     */
+    def stripNull(implicit ctx: Context): Type = {
+      assert(ctx.explicitNulls)
+      self.widenDealias.normNullableUnion match {
+        case OrType(lhs, rhs) if rhs.isNullType => lhs
+        case _ => self
+      }
+    }
+
+    /** Like `stripNull`, but removes only the `JavaNull`s. */
+    def stripJavaNull(implicit ctx: Context): Type = {
+      assert(ctx.explicitNulls)
+      self.widenDealias.normNullableUnion match {
+        case OrType(lhs, rhs) if rhs.isJavaNullType => lhs
+        case _ => self
+      }
+    }
+
+    /** Collapses all `JavaNull` unions within this type, and not just the outermost ones (as `stripJavaNull` does).
+     *  e.g. (Array[String|Null]|Null).stripNull => Array[String|Null]
+     *       (Array[String|Null]|Null).stripInnerNulls => Array[String]
+     *  If no `JavaNull` unions are found within the type, then returns the input type unchanged.
+     */
+    def stripAllJavaNull(implicit ctx: Context): Type = {
+      assert(ctx.explicitNulls)
+      var diff = false
+      object RemoveNulls extends TypeMap {
+        override def apply(tp: Type): Type =
+          tp.normNullableUnion match {
+            case OrType(lhs, rhs) if rhs.isJavaNullType =>
+              diff = true
+              mapOver(lhs)
+            case _ => mapOver(tp)
+          }
+      }
+      val rem = RemoveNulls(self.widenDealias)
```

Comment:
You can avoid the `diff` side effect be changing to:
```
  val self1 = self.widenDealias
  val rem = RemoveNulls(self1)
  if rem ne self1 then rem else self
```

Reply:
The variable is removed